### PR TITLE
docs: Add parser testing strategy discussion paper

### PR DESCRIPTION
**Pull request by Bernadette-Mohr**
- Covers unit, integration, and E2E testing principles
- Includes examples and best practices
- Documents testing schedule and anti-patterns

### CODE REVIEW

File: docs/parser_testing_strategy.md [hunk 1]
```diff
@@ -0,0 +1,813 @@
...
+* **Pinpoint failures: When a unit test fails, you know exactly which function broke**  
+* **Enable refactoring**: Safe to restructure code when comprehensive unit tests exist  
+* **Document behavior**: Unit tests serve as executable documentation of what each function should do  
+* **Catch logic errors: Off-by-one errors, incorrect conditionals, edge cases in algorithms**
+
+#### When to write
+
+* For any function with conditional logic (if/else, loops)  
+* For data transformation functions (parsing, formatting, validation)  
+* For calculations and algorithms
```
Comment (ndaelman-hu) [inline docs/parser_testing_strategy.md:32]: Is kind vague

File: docs/parser_testing_strategy.md [hunk 2]
```diff
@@ -0,0 +1,813 @@
...
+
+#### Purpose
+
+* **Fast feedback**: Developers get instant feedback during coding (\< 10ms execution)  
+* **Pinpoint failures: When a unit test fails, you know exactly which function broke**  
+* **Enable refactoring**: Safe to restructure code when comprehensive unit tests exist  
+* **Document behavior**: Unit tests serve as executable documentation of what each function should do  
+* **Catch logic errors: Off-by-one errors, incorrect conditionals, edge cases in algorithms**
+
+#### When to write
```
Comment (ndaelman-hu) [inline docs/parser_testing_strategy.md:28]: I would redo this section, as it is kinda hand-wavy. More important is that tests build of off expected formats or interfaces (enhanced with type hinting). Lastly, don't forget that Coveralls already provides metrics for how well our tests cover the repo.

**Review by ndaelman-hu — APPROVED** [review]
I already provided most of my feedback yesterday. I think the intro is new, so I just checked that.

**Comment by ndaelman-hu** [comment]
I suggest that we start focusing on applying the framework.

**Comment by JFRudzinski** [comment]
@ndaelman-hu What should we do with this? Is it still useful?

**Comment by ndaelman-hu** [comment]
Yes, still useful — I'd like to keep developing this rather than drop it. Overall I'm convinced we want a fairly wide testing platform, including some tiers we accept as more expensive to run. In an age of AI-written code the test harness becomes more load-bearing, so it's worth investing in.

I've got a few concrete pieces of feedback I'll already apply to the doc:
- The unit/integration/E2E level is presented as *the* framework. I'll reframe it as one dimension of testing. It's genuinely useful right now — we only write unit tests today, so laying out the levels makes the missing E2E tier obvious and helps surface forgotten test modes — but I want room to add other axes later (property-based / Hypothesis, etc.) without reworking it.
- The unit-test example asserts a feature of the NOMAD/library object it builds rather than of our own code, so it teaches the wrong thing. I'll replace or drop it.
- I'll also distill the actionable parts into `.github/copilot-instructions.md` — short and example-free, since Copilot needs the rules, not the walkthroughs.

@ladinesa, would appreciate your feedback on this as well.

**Comment by ladinesa** [comment]
> Yes, still useful — I'd like to keep developing this rather than drop it. Overall I'm convinced we want a fairly wide testing platform, including some tiers we accept as more expensive to run. In an age of AI-written code the test harness becomes more load-bearing, so it's worth investing in.
> 
> I've got a few concrete pieces of feedback I'll already apply to the doc:
> 
>     * The unit/integration/E2E level is presented as _the_ framework. I'll reframe it as one dimension of testing. It's genuinely useful right now — we only write unit tests today, so laying out the levels makes the missing E2E tier obvious and helps surface forgotten test modes — but I want room to add other axes later (property-based / Hypothesis, etc.) without reworking it.
> 
>     * The unit-test example asserts a feature of the NOMAD/library object it builds rather than of our own code, so it teaches the wrong thing. I'll replace or drop it.
> 
>     * I'll also distill the actionable parts into `.github/copilot-instructions.md` — short and example-free, since Copilot needs the rules, not the walkthroughs.
> 
> 
> @ladinesa, would appreciate your feedback on this as well.

I would align this with the current testing strategy we have instead of developing it independently. Most of the points here are already covered and would simply implement the ones not yet covered.

**Comment by ndaelman-hu** [comment]
> > Yes, still useful — I'd like to keep developing this rather than drop it. Overall I'm convinced we want a fairly wide testing platform, including some tiers we accept as more expensive to run. In an age of AI-written code the test harness becomes more load-bearing, so it's worth investing in.
> > I've got a few concrete pieces of feedback I'll already apply to the doc:
> > ```
> > * The unit/integration/E2E level is presented as _the_ framework. I'll reframe it as one dimension of testing. It's genuinely useful right now — we only write unit tests today, so laying out the levels makes the missing E2E tier obvious and helps surface forgotten test modes — but I want room to add other axes later (property-based / Hypothesis, etc.) without reworking it.
> > 
> > * The unit-test example asserts a feature of the NOMAD/library object it builds rather than of our own code, so it teaches the wrong thing. I'll replace or drop it.
> > 
> > * I'll also distill the actionable parts into `.github/copilot-instructions.md` — short and example-free, since Copilot needs the rules, not the walkthroughs.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @ladinesa, would appreciate your feedback on this as well.
> 
> I would align this with the current testing strategy we have instead of developing it independently. Most of the points here are already covered and would simply implement the ones not yet covered.

@ladines done

File: docs/parser_testing_strategy.md [hunk 3]
```diff
@@ -0,0 +1,820 @@
...
+
+Scope is not the only axis. A second is how inputs are chosen: the examples below are all *example-based* (hand-picked inputs with known outputs). Asserting invariants over inputs rather than fixed cases is already part of the protocol as *metamorphic* tests (for example, changing whitespace or the source unit rescales the result predictably); generalising these with *property-based* testing (e.g. Hypothesis, over generated inputs) is a natural next step. Treat the scope ladder below as one well-developed dimension of the strategy, not the whole of it.
+
+### Scope: different test types serve different purposes and catch different categories of bugs
+
+1. **Unit Tests** catch logic errors early and cheaply  
+2. **Integration Tests** catch interface mismatches and coordination issues  
+3. **E2E Tests** catch real-world edge cases and validate user-facing behavior
```
Comment (ladinesa) [inline docs/parser_testing_strategy.md]: large fixtures?

File: docs/parser_testing_strategy.md [hunk 4]
```diff
@@ -0,0 +1,820 @@
+## Fundamentals
```
Comment (ladinesa) [inline docs/parser_testing_strategy.md]: i thought we are supposed to extend how_to/test_parsers.md
